source/aws_imds_client.c

```diff
@@ -0,0 +1,953 @@
+        AWS_LOGF_ERROR(
+            AWS_LS_IMDS_CLIENT,
+            "id=%p: Error while recording successful retry: %s",
+            (void *)client,
+            aws_error_str(aws_last_error()));
+    }
+
+    s_query_complete(imds_user_data);
+}
+
+static void s_on_retry_token_acquired(
+    struct aws_retry_strategy *strategy,
+    int error_code,
+    struct aws_retry_token *token,
+    void *user_data) {
+    (void)strategy;
+
+    struct imds_user_data *imds_user_data = user_data;
+    struct aws_imds_client *client = imds_user_data->client;
+
+    if (!error_code) {
+        AWS_LOGF_WARN(AWS_LS_IMDS_CLIENT, "id=%p: IMDS Client successfully acquired retry token.", (void *)client);
+        imds_user_data->retry_token = token;
+        client->function_table->aws_http_connection_manager_acquire_connection(
+            client->connection_manager, s_on_acquire_connection, imds_user_data);
+    } else {
+        AWS_LOGF_WARN(
+            AWS_LS_IMDS_CLIENT,
+            "id=%p: IMDS Client failed to acquire retry token, error code %d(%s)",
+            (void *)client,
+            error_code,
+            aws_error_str(error_code));
+        imds_user_data->error_code = error_code;
+        s_query_complete(imds_user_data);
+    }
+}
+
+static enum imds_token_copy_result s_copy_token_safely(struct imds_user_data *user_data) {
+    struct aws_imds_client *client = user_data->client;
+    enum imds_token_copy_result ret = AWS_IMDS_TCR_UNEXPECTED_ERROR;
+
+    struct aws_linked_list pending_queries;
+    aws_linked_list_init(&pending_queries);
+    aws_mutex_lock(&client->token_lock);
+
+    if (client->token_state == AWS_IMDS_TS_VALID) {
+        aws_byte_buf_reset(&user_data->imds_token, true);
+        struct aws_byte_cursor cursor = aws_byte_cursor_from_buf(&client->cached_token);
+        if (aws_byte_buf_append_dynamic(&user_data->imds_token, &cursor)) {
+            ret = AWS_IMDS_TCR_UNEXPECTED_ERROR;
+        } else {
+            ret = AWS_IMDS_TCR_SUCCESS;
+        }
+    } else {
+        ret = AWS_IMDS_TCR_WAITING_IN_QUEUE;
+        struct imds_token_query *query = aws_mem_calloc(client->allocator, 1, sizeof(struct imds_token_query));
+        if (query != NULL) {
+            query->user_data = user_data;
+            aws_linked_list_push_back(&client->pending_queries, &query->node);
+        } else {
+            ret = AWS_IMDS_TCR_UNEXPECTED_ERROR;
+        }
+
+        if (client->token_state == AWS_IMDS_TS_INVALID) {
+            if (s_client_start_query_token(client)) {
+                ret = AWS_IMDS_TCR_UNEXPECTED_ERROR;
+                aws_linked_list_swap_contents(&pending_queries, &client->pending_queries);
+            } else {
+                client->token_state = AWS_IMDS_TS_UPDATE_IN_PROGRESS;
+            }
+        }
+    }
+    aws_mutex_unlock(&client->token_lock);
+
+    /* poll swapped out pending queries if there is any */
```

Comment:
Feels like completing queries could be a separate function that takes the list, error, token copy

Reply:
makes sense.